### PR TITLE
Mob stays vertical while in cryo

### LINKED ISSUE #54766: When in cryo, you go sideways.
## Round ID:
////

## Testmerges:
/////

## Reproduction:
Get into a cryo chamber and have a friendly doc turn it on for you.
Go to sleep, as you do in cryo.
You are now sideways cos you're sleeping inside cryo making it look funny.

### PR DESCRIPTION
**Pull request by prodirus**
## About The Pull Request

Fixes #54766 

Currently, mobs in cryo will turn horizontal in the cryotube when they fall unconscious. This should stop that.

## Why It's Good For The Game

No more funny beheaded and delimbed horizontal spaceman in cryotube.

## Changelog
:cl:
fix: New cryomix formulation ensures crewmember's bodies stay vertical during cryostasis!
/:cl:

### CODE REVIEW

File: code/modules/mob/living/init_signals.dm [hunk 1]
```diff
@@ -79,6 +79,8 @@
 	if(buckled && buckled.buckle_lying != NO_BUCKLE_LYING)
 		return // Handled by the buckle.
 	mobility_flags &= ~MOBILITY_STAND
+	if(HAS_TRAIT_FROM(src, TRAIT_IMMOBILIZED, CRYO_TRAIT))
+		return // Don't go horizontal if mob falls asleep from cryo while in cryotube
```
Comment (LemonInTheDark) [inline code/modules/mob/living/init_signals.dm:83]: Paging @Rohesie for this, this seems like a rather hacky way of preventing the issue

Comment (prodirus) [reply code/modules/mob/living/init_signals.dm:83]: If there is a better way, I'm all ears! 

From what I can tell cryo setting the mob to fall unconscious means it gets the TRAIT_FLOORED trait from set_stat. This then calls on_floored_trait_gain, which calls on_floored_start. This proc then makes the sprite rotate.

Comment (Rohesie) [reply code/modules/mob/living/init_signals.dm:83]: This is less than ideal.

Make cryo add and remove a more generic trait besides this one, something to signify the contrary of `FLOORED`. Maybe `FORCED_STANDING` or the likes. Ideally we should port Naksu's traits in Goon for this, but a normal trait should do it in the meantime.

Then you substitute this check for that one there. Do it before the mobility flag change, though, same as the buckled check.

Comment (prodirus) [reply code/modules/mob/living/init_signals.dm:83]: Alright, will take a look!

Comment (prodirus) [reply code/modules/mob/living/init_signals.dm]: @Rohesie Changes made, this closer to what should be done?

**Comment by LemonInTheDark** [comment]
Except this time it's actually quite good

**Review by Rohesie — CHANGES_REQUESTED** [review]
Okay, this is good.
Now search in the code for this:
```
RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_FLOORED), .proc/on_floored_trait_gain)
RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_FLOORED), .proc/on_floored_trait_loss)
```
You'll want to add to more of these, something like this:
```
RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_FORCED_STANDING), .proc/on_forced_standing_trait_gain)
RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_FORCED_STANDING), .proc/on_forced_standing_trait_loss)
```
Also the procs, like `on_floored_trait_gain` and `on_floored_trait_loss`

In `on_forced_standing_trait_loss` you'll want to add this part:
```
if(occupant.resting || HAS_TRAIT(occupant, TRAIT_FLOORED))
	occupant.set_lying_down()
```

And in `on_forced_standing_trait_gain` you'll want to add this part:
```
occupant.set_body_position(STANDING_UP)
occupant.set_lying_angle(0)
```

Poke me on discord if you need help, I can guide you through.

**Comment by prodirus** [comment]
> Okay, this is good.
> Now search in the code for this:
> 
> ```
> RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_FLOORED), .proc/on_floored_trait_gain)
> RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_FLOORED), .proc/on_floored_trait_loss)
> ```
> 
> You'll want to add to more of these, something like this:
> 
> ```
> RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_FORCED_STANDING), .proc/on_forced_standing_trait_gain)
> RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_FORCED_STANDING), .proc/on_forced_standing_trait_loss)
> ```
> 
> Also the procs, like `on_floored_trait_gain` and `on_floored_trait_loss`
> 
> In `on_forced_standing_trait_loss` you'll want to add this part:
> 
> ```
> if(occupant.resting || HAS_TRAIT(occupant, TRAIT_FLOORED))
> 	occupant.set_lying_down()
> ```
> 
> And in `on_forced_standing_trait_gain` you'll want to add this part:
> 
> ```
> occupant.set_body_position(STANDING_UP)
> occupant.set_lying_angle(0)
> ```
> 
> Poke me on discord if you need help, I can guide you through.

Done.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 2]
```diff
@@ -41,6 +42,8 @@
 	vis_contents += occupant
 	pixel_y = 22
 	ADD_TRAIT(occupant, TRAIT_IMMOBILIZED, CRYO_TRAIT)
+	// Keep them standing! They'll go sideways in the tube when they fall asleep otherwise.
+	ADD_TRAIT(occupant, TRAIT_FORCED_STANDING, CRYO_TRAIT)
 	occupant.set_body_position(STANDING_UP)
 	occupant.set_lying_angle(0)
```
Comment (Rohesie) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:48]:
```suggestion
	ADD_TRAIT(occupant, TRAIT_FORCED_STANDING, CRYO_TRAIT)
```
These are now already processed when the trait is added for the first time.

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm [hunk 3]
```diff
@@ -30,6 +30,7 @@
 	if(occupant)
 		vis_contents -= occupant
 		REMOVE_TRAIT(occupant, TRAIT_IMMOBILIZED, CRYO_TRAIT)
+		REMOVE_TRAIT(occupant, TRAIT_FORCED_STANDING, CRYO_TRAIT)
 		if(occupant.resting || HAS_TRAIT(occupant, TRAIT_FLOORED))
 			occupant.set_lying_down()
```
Comment (Rohesie) [inline code/modules/atmospherics/machinery/components/unary_devices/cryo.dm:35]:
```suggestion
		REMOVE_TRAIT(occupant, TRAIT_FORCED_STANDING, CRYO_TRAIT)
```
And this one is processed when the trait is properly removed.

**Review by Rohesie — CHANGES_REQUESTED** [review]
The check now needs to be added to `/mob/living/proc/set_resting`

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 2 above)
Comment (prodirus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: Done!

File: code/modules/atmospherics/machinery/components/unary_devices/cryo.dm (hunk 3 above)
Comment (prodirus) [reply code/modules/atmospherics/machinery/components/unary_devices/cryo.dm]: ....and done!

**Pull request merged**